### PR TITLE
Improvements to various DBs compatibility, serialization of JdbcMetaData to/from JSON, small changes to improve lineage

### PR DESCRIPTION
**Pull request by dpavlis**
Changes divided to two categories:

1. existing functionality extensions/changes:

- enhancements to support multiple DB implementations (MS SQL/MySQL/PostgreSQL/Oracle/Snowflake) when extracting metadata through JDBC connection
- improvement in reporting scope for ParenthesedSelect

2. additional functionality

- serialization of JdbcMetaData to/from JSON to allow "off-line" analysis of SQLs - added dependency on com.fasterxml.jackson
- implementing minimized variant of JsonTreeBuilder lineage output - better for sharing through API

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1,3 +1,7 @@
+# *Changes implemented to the original JSQLTranspiler*
```
Comment (manticore-projects) [inline README.md:1]: We don't need this in the Readme, instead it will appear in the Changelog generated by Bit automatically

File: build.gradle [hunk 2]
```diff
@@ -60,7 +60,7 @@ dependencies {
     // for the ASCII Trees
     testImplementation 'hu.webarticum:tree-printer:+'
 
-    testImplementation 'org.duckdb:duckdb_jdbc:1.0.0'
+    testImplementation 'org.duckdb:duckdb_jdbc:+'
```
Comment (manticore-projects) [inline build.gradle:63]: This is correct in general, however DUCKDB 1.1.1 introduced a CSV related bug which affects our test suite.
DuckDB fixed that already but we would need to wait for the next release.

File: build.gradle [hunk 3]
```diff
@@ -69,6 +69,16 @@ dependencies {
 
     // Java Doc in XML Format
     xmlDoclet 'com.manticore-projects.tools:xml-doclet:+'
+
```
Comment (manticore-projects) [inline build.gradle:72]: I am familiar with Jackson and use it myself for serialization.
But I don't like it here because it depends a massive 3rd party dependency for rather limited benefit:

1) the serialization is for caching only
2) the meta data are fairly simple and not nested

So I would like to ask for either writing the JSON manually (if you really need/insist in JSON serialisation) or just your the standard Java serialisation (since its for caching only)

File: src/main/java/ai/starlake/transpiler/JSQLColumResolver.java [hunk 4]
```diff
@@ -194,6 +195,28 @@ public JdbcResultSetMetaData getResultSetMetaData(String sqlStr) throws JSQLPars
     }
   }
 
+  /**
```
Comment (manticore-projects) [inline src/main/java/ai/starlake/transpiler/JSQLColumResolver.java:198]: Please why do we need this method here?

File: src/main/java/ai/starlake/transpiler/JSQLColumResolver.java [hunk 5]
```diff
@@ -266,6 +289,21 @@ public <T> T getLineage(Class<? extends TreeBuilder<T>> treeBuilderClass, String
     return builder.getConvertedTree(this);
   }
 
+
```
Comment (manticore-projects) [inline src/main/java/ai/starlake/transpiler/JSQLColumResolver.java:292]: Please why do we need this method here?

File: src/main/java/ai/starlake/transpiler/schema/JdbcCatalog.java [hunk 6]
```diff
@@ -42,21 +51,25 @@ public JdbcCatalog(String tableCatalog, String catalogSeparator) {
     this.catalogSeparator = catalogSeparator != null ? catalogSeparator : ".";
   }
 
+  public JdbcCatalog() {
```
Comment (manticore-projects) [inline src/main/java/ai/starlake/transpiler/schema/JdbcCatalog.java:54]: this should not be needed

File: src/main/java/ai/starlake/transpiler/schema/JdbcColumn.java [hunk 7]
```diff
@@ -48,6 +58,7 @@ public class JdbcColumn implements Comparable<JdbcColumn> {
   public String scopeCatalog;
   public String scopeSchema;
   public String scopeTable;
+  public String scopeColumn;
```
Comment (manticore-projects) [inline src/main/java/ai/starlake/transpiler/schema/JdbcColumn.java:61]: Why is this needed? Can you provide a simple example for illustration please where the column name is different from scope column name? In my understanding this is what Alias is used for?

File: src/main/java/ai/starlake/transpiler/schema/JdbcColumn.java [hunk 8]
```diff
@@ -131,31 +144,36 @@ public JdbcColumn(String tableCatalog, String tableSchema, String tableName, Str
...
   public JdbcColumn(String columnName, Expression expression) {
     this("", "", "", columnName, Types.OTHER, "Other", 0, 0, 10, columnNullableUnknown, "", "", 0,
-        0, "", "", "", "", (short) 0, "", "", expression);
+        0, "", "", "", "", "", (short) 0, "", "", expression);
   }
 
   public JdbcColumn(String columnName) {
     this(columnName, new Column(columnName));
   }
+
```
Comment (manticore-projects) [inline src/main/java/ai/starlake/transpiler/schema/JdbcColumn.java:172]: Again, should not be needed.

File: src/main/java/ai/starlake/transpiler/schema/JdbcMetaData.java [hunk 9]
```diff
@@ -41,31 +44,48 @@
...
+@JsonIgnoreProperties(ignoreUnknown = true)
 @SuppressWarnings({"PMD.CyclomaticComplexity"})
 public final class JdbcMetaData implements DatabaseMetaData {
   public final static Logger LOGGER = Logger.getLogger(JdbcMetaData.class.getName());
   public static final Map<Integer, String> SQL_TYPE_NAME_MAP = new HashMap<>();
 
-  private final CaseInsensitiveLinkedHashMap<JdbcCatalog> catalogs =
+  private CaseInsensitiveLinkedHashMap<JdbcCatalog> catalogs =
       new CaseInsensitiveLinkedHashMap<>();
-  private final String currentCatalogName;
```
Comment (manticore-projects) [inline src/main/java/ai/starlake/transpiler/schema/JdbcMetaData.java:54]: I think this better shall be final. We are not sure what happens when you change those during runtime/lifetime of the object. I would prefer rather immutable objects.

File: src/main/java/ai/starlake/transpiler/schema/JdbcMetaData.java [hunk 10]
```diff
@@ -164,14 +184,16 @@ public JdbcMetaData() {
    * @throws SQLException when the database fails to return CURRENT_CATALOG or CURRENT_SCHEMA
    */
   public JdbcMetaData(Connection con) throws SQLException {
-    // todo: customise this for various databases, e. g. Oracle would need a "FROM DUAL"
+	DatabaseMetaData metaData = con.getMetaData();
+	this.databaseType = JdbcUtils.DatabaseSpecific.getType(metaData.getDatabaseProductName());
+	
     try (Statement statement = con.createStatement();
-        ResultSet rs = statement.executeQuery("SELECT current_database(), current_schema()")) {
+        ResultSet rs = statement.executeQuery(this.databaseType.getCurrentSchemaQuery())) {
```
Comment (manticore-projects) [inline src/main/java/ai/starlake/transpiler/schema/JdbcMetaData.java:191]: You have a strong point here, well done!

File: src/main/java/ai/starlake/transpiler/schema/JdbcSchema.java [hunk 11]
```diff
@@ -43,23 +52,28 @@ public JdbcSchema(String tableSchema, String tableCatalog) {
     this.tableCatalog = tableCatalog != null ? tableCatalog : "";
   }
 
+  public JdbcSchema() {
```
Comment (manticore-projects) [inline src/main/java/ai/starlake/transpiler/schema/JdbcSchema.java:55]: Should not be necessary

File: src/main/java/ai/starlake/transpiler/schema/JdbcUtils.java [hunk 12]
```diff
@@ -0,0 +1,154 @@
...
+import java.util.List;
+
+public class JdbcUtils {
+
+
+	/**
+	 * Used for detecting RDBMS type and DB specific handling 
+	 */
+	public enum DatabaseSpecific {
+		  ORACLE("ORACLE",new String[] {"SYNONYM","TABLE","VIEW"}, new String[] {"SYS"}, "SELECT SYS_CONTEXT('USERENV', 'DB_NAME') AS database_name , SYS_CONTEXT('USERENV', 'CURRENT_SCHEMA') AS current_schema FROM dual"),
```
Comment (manticore-projects) [inline src/main/java/ai/starlake/transpiler/schema/JdbcUtils.java:18]: This is good stuff, I like it!

**Review by manticore-projects — CHANGES_REQUESTED** [review]
Thank you so much for your interest, work and contribution! It is much appreciated.
I have 2 main concerns though:
- I don't like the Jackson dependency, because it is rather large/fat for minimal benefit
- I am not sure about the new `scopeColumn` but maybe I miss something

Both points can be negotiated of course and I am not dogmatic here.
@hayssams: what is your opinion on pulling Jackson XML in please?

**Comment by manticore-projects** [comment]
Please always run "gradle verify" before submitting to ensure that CI/QA succeeds.

File: src/main/java/ai/starlake/transpiler/JSQLColumResolver.java (hunk 5 above)
Comment (dpavlis) [reply src/main/java/ai/starlake/transpiler/JSQLColumResolver.java:292]: Ability to configure the underlying parser - e.g. to handle T-SQL way of quoting (square brackets)

File: src/main/java/ai/starlake/transpiler/JSQLColumResolver.java (hunk 4 above)
Comment (dpavlis) [reply src/main/java/ai/starlake/transpiler/JSQLColumResolver.java:198]: Ability to configure the underlying parser - e.g. to handle T-SQL way of quoting (square brackets).

File: README.md (hunk 1 above)
Comment (dpavlis) [reply README.md:1]: sure.. commited by mistake.

File: build.gradle (hunk 3 above)
Comment (dpavlis) [reply build.gradle:72]: Well, this is one of the core extensions of existing functionality for my usecase (and I believe others would benefit from it too). Writing the serialization as a series of append() is possible, but I consider that to be an ugly way of doing the serialization - not mentioning the de-serialization back. I could implement it as an external class, but then would need additional public methods added (or change too many members to be public - which, btw, the public visibility widely used my not be the best thing).

File: src/main/java/ai/starlake/transpiler/schema/JdbcColumn.java (hunk 7 above)
Comment (dpavlis) [reply src/main/java/ai/starlake/transpiler/schema/JdbcColumn.java:61]: In case of SQL statements like this:
select * from a, (select cola id, colb b1 from mytable) b where a.id=b.id

the resulting lineage in scope attribute/value references the renamed columns (they are copied into the resultset during analysis) adding scopeColumn allows to preserve the orginal name thus providing the correct (for my needs - the original source of data) scope information.

File: src/main/java/ai/starlake/transpiler/schema/JdbcColumn.java (hunk 8 above)
Comment (dpavlis) [reply src/main/java/ai/starlake/transpiler/schema/JdbcColumn.java:172]: as explained above

File: src/main/java/ai/starlake/transpiler/schema/JdbcMetaData.java (hunk 9 above)
Comment (dpavlis) [reply src/main/java/ai/starlake/transpiler/schema/JdbcMetaData.java:54]: The `LinkedHashMap catalogs` could remain final, `currentCatalogName` not - due to the added (de)serialization. 
Plus, possibly supporting changing of current catalog & current schema could support doing the statement analysis under different context on larger DB catalogs where there are multiple catalogs and/or schemas where tables&columns referenced by SQL query can be different.
This again supports the idea that the DB catalog is extracted once and then potentially analyzed multiple times with different context set.

**Comment by hayssams** [comment]
> Thank you so much for your interest, work and contribution! It is much appreciated. I have 2 main concerns though:
> 
> * I don't like the Jackson dependency, because it is rather large/fat for minimal benefit
> * I am not sure about the new `scopeColumn` but maybe I miss something
> 
> Both points can be negotiated of course and I am not dogmatic here. @hayssams: what is your opinion on pulling Jackson XML in please?

I agree that Jackson may be too overkill to for this minimal task. 
We may also face issues in projects that used different versions of Jackson.
We usually rely on Jackson only if the source project already rely on it (aka Spark).

I suggest that we rely on the lightweight org.json.json library

**Comment by hayssams** [comment]
regarding the XML part of Jackson, doesn't Java has its own marshaller (JAXB - jakarta.xml.bind) which would put us on the safe side. 
Still not a fan of it but IMHO this would be better than bringing in Jackson-XML.
Overall, based on our data structure, we even could implement both (json & XML) through reflection

**Comment by manticore-projects** [comment]
Suggest way forward to a merge: 
- remove the Jackson dependency
- pass the QA/Ci tests via `gradle verify`
- merge

**Comment by dpavlis** [comment]
OK, I have re-implemented the JSON serialization - got rid of Jackson.

**Comment by manticore-projects** [comment]
Thank you again for your interest and contribution. I merge it although I would like to ask again for  running `gradle verify` before submitting. I will sort the QA/Ci myself now. Cheers!

**Comment by manticore-projects** [comment]
I am sorry, but your PR breaks all the TickitDB related tests and I have not been able to find out the reason in a hurry.
So please, as a way forward:

    gradle check
    fix QA/CI exceptions
    ensure all tests are still running and fix where needed

(Only then) I can merge again.

**Comment by dpavlis** [comment]
OK. so I added DatabaseSpecific for DuckDB to correctly get catalog & schema (that was the reasons for some tests to fail with "can't find schema/column". All tests now pass. So should be ready for inclusion ;-)

**Comment by manticore-projects** [comment]
Please resubmit your PR, thank you!
